### ISSUE TITLE
test-echo.c does not work on localhost

### ISSUE THREAD

**Issue by namowen** [warmcat/libwebsockets#716]
test-echo works properly with:
libwebsockets-test-echo --client echo.websocket.org

but using with localhost against a local test-echo server, nothing works:
libwebsockets-test-echo --client localhost --port 7681

I nailed it down to lws_callback_on_writable_all_protocol, which does not work. In pollfd.c, lws_callback_on_writable_all_protocol_vhost, the line
wsi = vhost->same_vh_protocol_list[protocol - vhost->protocols];
sets wsi to 0, so the while(wsi) block is never executed.

Replacing lws_callback_on_writable_all_protocol with lws_callback_on_writable(wsi) in test-echo.c main(), everything works fine.

Somehow strange...

Using libwebsockets version 2.0.1, ubuntu 16.04 64bit

**Comment by lws-team** [comment]
I'll look at it tomorrow, but my first guess would be it's related to it using NULL ws protocol (for compatability with echo.websocket.org) somehow.

**Comment by lws-team** [comment]
Yes... it was a good guess... I pushed a fix on v2.1-stable and master.

Did you mean that you are on v2.1?  The corresponding code with the problem doesn't exist on v2.0[.x]

**Comment by namowen** [comment]
It's 2.0.1 (see below). I downloaded it as tarball from github end of may 2016 to upgrade my old 1.3 version.
But I will upgrade to latest official stable version (using git this time) in the next weeks ;-)

####:/usr/local/bin$ libwebsockets-test-echo --client echo.websocket.org
[2016/11/15 10:10:14:8919] NOTICE: Built to support client operations
[2016/11/15 10:10:14:8920] NOTICE: Built to support server operations
lwsts[35403]: libwebsockets test server echo - license LGPL2.1+SLE
lwsts[35403]: (C) Copyright 2010-2016 Andy Green <andy@warmcat.com>
lwsts[35403]: Running in client mode
lwsts[35403]: Initial logging level 7
lwsts[35403]: **Libwebsockets version: 2.0.1** unknown-build-hash
lwsts[35403]: IPV6 compiled in and enabled
lwsts[35403]: libev support not compiled in
lwsts[35403]: libuv support not compiled in
lwsts[35403]:  Threads: 1 each 1024 fds
lwsts[35403]:  mem: platform fd map:  8192 bytes
lwsts[35403]:  Compiled with OpenSSL support
lwsts[35403]:  SSL disabled: no LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT
lwsts[35403]: Creating Vhost 'default' port -1, 1 protocols
lwsts[35403]:  mem: per-conn:          488 bytes + protocol rx buf
lwsts[35403]:  canonical_hostname = iotalive1
lwsts[35403]: Client connecting to echo.websocket.org:80....
lwsts[35403]: lws_protocol_init
lwsts[35403]: Client TX: hello from libwebsockets-test-echo client pid 35403 index 0
lwsts[35403]: Client RX: hello from libwebsockets-test-echo client pid 35403 index 0

**Comment by lws-team** [comment]
The tag `v2.0.1` is a "fly in amber", it can't be patched or changed.

I checked v2.0-stable, which is the maintained v2.0 branch with fixes, echo works fine on that as of today.

v2.0.1 is a long way behind v2.0-stable including some important security updates.  you should definitely update it, even if only to current v2.0-stable.

So there's nothing further for me to do here I think.

**Comment by namowen** [comment]
Sorry to open again.
Using now v2.1-stable, there is a bug in test-echo.c, running as server. If it receives the first text to echo from the client, pss is not initialized (actually all members are 0). Therefore on
line 109
assert((int)pss->len == -1);
hits and the server aborts.
Removing this line or adding LWS_CALLBACK_ESTABLISHED to initialize pss (as in LWS_CALLBACK_CLIENT_ESTABLISHED) would fix it.
Sorry that I didn't fix it, but I don't know (yet) how to create and push a patch with git.

tested fix:
after line 67: #ifndef LWS_NO_SERVER just insert

	case LWS_CALLBACK_ESTABLISHED:
		pss->index = 0;
		pss->len = -1;
		break;

**Comment by lws-team** [comment]
Thanks... I pushed the ESTABLISHED fix on v1.7 thru master.

About patches, because it started off as a git clone, you can get a "patch" of your changes without any effort, just type `git diff`.  You can then remove the parts that aren't relevant and cut-and-paste it as a first step... that's still a valid diff that can be used with `patch` if you trimmed it carefully.

A "patch" is just that with some lines like this above it

```
From: name <your@email.com>

One line patch description

(optional longer patch description)

```

**Comment by namowen** [comment]
Thx a lot. I will try to get into git a little bit more when (and if) I have time over Xmas :-)
